module/VuFind/src/VuFind/AjaxHandler/GetSearchResults.php

```diff
@@ -0,0 +1,455 @@
+        return $this->formatResponse(compact('elements'));
+    }
+
+    /**
+     * Get search results
+     *
+     * @param Params $params Request params
+     *
+     * @return ?Results
+     */
+    protected function getSearchResults(Params $params): ?Results
+    {
+        parse_str($params->fromQuery('querystring', ''), $searchParams);
+        $backend = $params->fromQuery('source', DEFAULT_SEARCH_BACKEND);
+        $searchType = $params->fromQuery('searchType', '');
+        if ('versions' === $searchType) {
+            $driverAndKeys = $this->versionsHelper->getDriverAndWorkKeysFromParams(
+                $searchParams,
+                $backend
+            );
+            if (empty($driverAndKeys['keys'])) {
+                return null;
+            }
+
+            $searchParams['lookfor'] = $this->versionsHelper
+                ->getSearchStringFromWorkKeys((array)$driverAndKeys['keys']);
+            $searchParams['type'] = $this->versionsHelper->getWorkKeysSearchType();
+        }
+
+        $results = $this->resultsManager->get($backend);
+        $paramsObj = $results->getParams();
+        $paramsObj->getOptions()->spellcheckEnabled(false);
+        $paramsObj->initFromRequest(new Parameters($searchParams));
+
+        if ('versions' !== $searchType) {
+            $this->saveSearchToHistory($results);
+        }
+
+        return $results;
+    }
+
+    /**
+     * Render page elements
+     *
+     * @param Params  $params  Request params
+     * @param Results $results Search results
+     *
+     * @return array
+     */
+    protected function getElements(Params $params, Results $results): array
+    {
+        if ($params->fromQuery('searchType', '') === 'versions') {
+            // Customize the URL helper to make sure it builds proper versions URLs
+            $results->getUrlQuery()
+                ->setDefaultParameter('id', $params->fromQuery('id'))
+                ->setDefaultParameter('keys', $params->fromQuery('keys'))
+                ->setSuppressQuery(true);
+        }
+
+        $result = [];
+        foreach ($this->elements as $selector => $element) {
+            $content = call_user_func([$this, $element['method']], $params, $results);
+            if (null !== $content) {
+                $result[$selector] = [
+                    'content' => $content,
+                    'target' => $element['target'],
+                    'attrs' => $element['attrs'] ?? [],
+                ];
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * Render search results
+     *
+     * @param Params  $params  Request params
+     * @param Results $results Search results
+     *
+     * @return ?string
+     */
+    protected function renderResults(Params $params, Results $results): ?string
+    {
+        [$baseAction] = explode('-', $results->getOptions()->getSearchAction());
+        $templatePath = "$baseAction/results-list.phtml";
```

Comment:
Is it a problem that this is hard-coded to list view? What happens in grid view? Obviously I don't object to starting with the most commonly-used case, but I just want to be sure other views aren't broken, and I'm curious whether there are future plans (or should at least be a JIRA ticket) to support AJAX grids.

Reply:
results-list.phtml is the main template for the results regardless of the list type. Grid is already supported. :) Visual is not AJAX-enabled, but it doesn't need paging anyway.